client/go/outline/vpn/nmconn_linux.go

```diff
@@ -0,0 +1,188 @@
+
+	dev, err := c.waitForDeviceToBeAvailable()
+	if err != nil {
+		return errSetupVPN(nmLogPfx, "failed to find TUN device", err, "tun", c.nmOpts.TUNName)
+	}
+	slog.Debug(nmLogPfx+"located TUN device", "tun", c.nmOpts.TUNName, "dev", dev.GetPath())
+
+	if err = dev.SetPropertyManaged(true); err != nil {
+		return errSetupVPN(nmLogPfx, "failed to set TUN device to be managed", err, "dev", dev.GetPath())
+	}
+	slog.Debug(nmLogPfx+"set TUN device to be managed", "dev", dev.GetPath())
+
+	props := make(map[string]map[string]interface{})
+	configureCommonProps(props, c.nmOpts)
+	configureTUNProps(props)
+	configureIPv4Props(props, c.nmOpts)
+	slog.Debug(nmLogPfx+"populated NetworkManager connection settings", "settings", props)
+
+	// The previous SetPropertyManaged call needs some time to take effect
+	for retries := 10; retries > 0; retries-- {
+		slog.Debug(nmLogPfx+"trying to create connection for TUN device ...", "dev", dev.GetPath())
+		c.ac, err = c.nm.AddAndActivateConnection(props, dev)
+		if err == nil {
+			break
+		}
+		slog.Debug(nmLogPfx+"waiting for TUN device being managed", "err", err)
+		time.Sleep(50 * time.Millisecond)
+	}
+	if err != nil {
+		return errSetupVPN(nmLogPfx, "failed to create new connection for device", err, "dev", dev.GetPath())
+	}
+	slog.Info(nmLogPfx+"successfully configured NetworkManager connection", "conn", c.ac.GetPath())
+	return nil
+}
+
+func (c *linuxVPNConn) closeNMConnection() error {
+	if c == nil || c.nm == nil {
+		return nil
+	}
+
+	if c.ac != nil {
+		if err := c.nm.DeactivateConnection(c.ac); err != nil {
+			slog.Warn(nmLogPfx+"not able to deactivate connection", "err", err, "conn", c.ac.GetPath())
+		}
+		slog.Debug(nmLogPfx+"deactivated connection", "conn", c.ac.GetPath())
+
+		conn, err := c.ac.GetPropertyConnection()
+		if err == nil {
+			err = conn.Delete()
+		}
+		if err != nil {
+			return errCloseVPN(nmLogPfx, "failed to delete connection", err, "conn", c.ac.GetPath())
+		}
+		slog.Info(nmLogPfx+"connection deleted", "conn", c.ac.GetPath())
+	}
+
+	return nil
+}
+
+func (c *linuxVPNConn) waitForDeviceToBeAvailable() (dev gonm.Device, err error) {
+	for retries := 20; retries > 0; retries-- {
+		slog.Debug(nmLogPfx+"trying to find TUN device ...", "tun", c.nmOpts.TUNName)
+		dev, err = c.nm.GetDeviceByIpIface(c.nmOpts.TUNName)
+		if dev != nil && err == nil {
+			return
+		}
+		slog.Debug(nmLogPfx+"waiting for TUN device to be available", "err", err)
+		time.Sleep(50 * time.Millisecond)
+	}
+	return nil, errSetupVPN(nmLogPfx, "failed to find TUN device", err, "tun", c.nmOpts.TUNName)
+}
+
+func configureCommonProps(props map[string]map[string]interface{}, opts *nmConnectionOptions) {
+	props["connection"] = map[string]interface{}{
+		"id":             opts.Name,
+		"interface-name": opts.TUNName,
+	}
+}
+
+func configureTUNProps(props map[string]map[string]interface{}) {
+	props["tun"] = map[string]interface{}{
+		// The operating mode of the virtual device.
```

Comment:
It would be nice to have references to the documentation where this is specified.

Reply:
Done.